### PR TITLE
model skeleton and docstring

### PR DESCRIPTION
**Pull request by VictorGillioz**
This is the skeleton for a MeshGraphNet network that requires further implementation of an encoder, processor and decoder. The forward method takes as input a list of Graph objects, which does not allow efficient parallelism. However, with Graphs of varying node/edge sizes it might not be possible to efficiently parallelise computation. Moreover, there is an important chance that we will use batch size = 1.

### CODE REVIEW

File: model/model.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+from meshnets.datatypes.graph import Graph
+from meshnets.model.encoder import GraphEncoder
+from meshnets.model.decoder import GraphDecoder
+from meshnets.model.processor import GraphProcessor
+
+
+class MeshGraphNet(torch.nn.Module):
+    """MeshGraphNet model.
+    
+    It is composed of a GraphEncoder, a GraphProcessor and a GraphDecoder.
```
Comment (IvanPombo) [inline model/model.py:15]: Suggestion: "The model architecture is composed of graph Encoder, Processor and Decoder.
The input is a Graph object which is encoded together with its features to a latent representation. Thereafter, message passing steps are applied by the Processor on the latent Graph object and, at the end, the Decoder outputs features at each node."

File: model/model.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    It takes as input Graph objects, encodes their node and mesh features
+    to a latent representation, apply message passing steps
+    on the latent Graph objects, and decodes the output features
+    at each node."""
+
+    def __init__(self, node_feats_size, mesh_feats_size, output_size,
```
Comment (IvanPombo) [inline model/model.py:21]: What do you mean by "feats" it is to reference the features?

Do not be afraid to be as descriptive as possible. "node_features_size" is a wonderful name for the variable and a great way to document what you are doing.

Is the `num_mlp_layers` equal for both the edges and nodes MLP encoder?

File: model/model.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        super().__init__()
+
+        self.encoder = GraphEncoder(node_feats_size, mesh_feats_size,
+                                    latent_size, num_mlp_layers)
+
+        self.processor = GraphProcessor(latent_size, message_passing_steps)
+
+        self.decoder = GraphDecoder(output_size, latent_size, num_mlp_layers)
+
+    def forward(self, graph_list: np.ndarray[Graph]) -> torch.Tensor:
```
Comment (IvanPombo) [inline model/model.py:34]: Here, you are passing the input is an n-dimensional array of Graphs. Maybe, just it can just be `List[Graph]`, wdyt?

See [here](https://docs.python.org/3.6/library/typing.html).

**Review by IvanPombo — COMMENTED** [review]
It is taking shape! Left you minor things to discuss and iterate first before approving.

File: model/model.py (hunk 2 above)
Comment (VictorGillioz) [reply model/model.py:21]: Yes feats refers to features. Thanks for the suggestion, I will make the arguments more explicit.

From the original implementation, num_mlp_layers is identical for both edges and nodes MLP encoder, as well as for the decoder. The latent size is also identical for the edge and node features.

File: model/model.py (hunk 3 above)
Comment (VictorGillioz) [reply model/model.py:34]: A list probably makes more sense indeed as we don't need array properties here. I will change that!

**Comment by sarmento** [comment]
@VictorGillioz : I am adding @AugustoPeres who is developing some infra-structure that will help you run experiments at scale.

@AugustoPeres : pls have a look at what Victor is building. I think there is opportunity to start integrating part of this in the infra and use the infra here.

**Review by AugustoPeres — COMMENTED** [review]
Left a few comments. Which mainly relate to the following, more high level comment:

Why are we not using `pytorch_geometric` to work with graph data. It seems that there are several things that are not necessayr. Although I may be missing something.

For once, why do we have a custom `Graph` class implementation? I took a look at the `data_preprocessing` file and it seems that everything there could be done without this. For example, the function:

```python
def triangle_mesh_to_graph(nodes: np.ndarray, node_features: np.ndarray,
                           faces: np.ndarray) -> Graph:
    """Produce a Graph object from triangle mesh data.

    Triangle edges exist twice in the Graph, once in each direction.
    """

    senders, receivers = _triangles_to_edges(faces)
    mesh_features = _compute_mesh_features(nodes, senders, receivers)
    edge_set = EdgeSet(mesh_features, senders, receivers)

    graph = Graph(node_features, edge_set)

    return graph
```

could instead be:

```python
from pytorch_geometric import data
def triangle_mesh_to_graph(nodes: np.ndarray, node_features: np.ndarray,
                           faces: np.ndarray) -> Data:
    """Produce a Graph object from triangle mesh data.

    Triangle edges exist twice in the Graph, once in each direction.
    """

    senders, receivers = _triangles_to_edges(faces)
    mesh_features = _compute_mesh_features(nodes, senders, receivers)
    edge_set = EdgeSet(mesh_features, senders, receivers)

    graph = Data(x=node_features, edge_index=edge_set)

    return graph
```

Take a look at the [official documentation](https://pytorch-geometric.readthedocs.io/en/latest/get_started/introduction.html). Additionally, the usage of this library, will make [mini-batching](https://pytorch-geometric.readthedocs.io/en/latest/get_started/introduction.html#mini-batches) very easy.

Another advantage is that we could very easily pre-process the dataset, create the graphs, and save them to disk by just using pytorch by simply doing:

```python
graph = triangle_mesh_to_graph(args)
torch.save(graph, 'path')
```

Loading this graph would then simply be: `graph = torch.load('path')`

Also, I think that we may run into problems later dataloading as pytorch datasets expect specific classes to como out of the `DataSet` class. For example, we cannot have a pytorch dataset returning `string` obejcts and have it work with dataloader. Therefore I am not sure if we can create a `Dataset`that outputs `Graph`.

File: meshnets/model/model.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
+"""Define the main MeshGraphNet model"""
+
+import numpy as np
+import torch
+
+from meshnets.datatypes.graph import Graph
+from meshnets.model.encoder import GraphEncoder
+from meshnets.model.decoder import GraphDecoder
+from meshnets.model.processor import GraphProcessor
```
[the remark this answers is not in the document]
Comment (AugustoPeres) [reply meshnets/model/model.py:9]: Also where are the `GraphEncoder` and pther models implemented?

**Comment by AugustoPeres** [comment]
At the moment I am having a hard time understanding several choices:

1. The Graph, class implementation. It seems that we would be better served by using the pytorch class https://pytorch-geometric.readthedocs.io/en/latest/get_started/introduction.html#data-handling-of-graphs.
2. Why the model receives a list of graphs. This is not standard across any library. Models should always receive batched data in the form of tensors.

I would also like to take this opportunity to shamelessly promote today's Deep Learning Sessions Event which will cover exactly these topics. First we will look at how to represented graph data using pytorch. And then we will see how to write a graph neural network using pytorch.

https://www.meetup.com/deep-learning-sessions-lisboa/events/293365425/

File: meshnets/model/model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (AugustoPeres) [reply meshnets/model/model.py:9]: I already saw that they are not implemented. That was my fault. However when something is not implemented we shoudl replace the import with:

```python
class GraphEncoder:
    pass
```

**Comment by VictorGillioz** [comment]
Thanks for the extensive feedback @AugustoPeres! I am not very familiar with graph data so I really appreciate it. I will go through your comments and make global changes based on your suggestions.

**Comment by AugustoPeres** [comment]
> Thanks for the extensive feedback @AugustoPeres! I am not very familiar with graph data so I really appreciate it. I will go through your comments and make global changes based on your suggestions.

No problem. Although I advice you not to address a lot of these changes at the moment to prevent this pull request from becoming too large. There are several good changes that should be merged as soon as possible. For example, new file structure and repository organization.

There are comments that are important to address before merging such as the imports to the same file (https://github.com/inductiva/meshnets/pull/3#discussion_r1197620494).

However, the comments about the models and the `Data` class should be left for future refactoring. Better still if each one of those is its own separate pull request. Therefore what I advise is the following:

1. Address [this comment](https://github.com/inductiva/meshnets/pull/3#discussion_r1197620494) and the comments of the other reviewers;
2. Look into pytorch's `Data` examples for graphs and then make a refactoring where your implementation is replaced by their class;
3. Look at `torch_geometric` layers and make another refactoring where you implement the necessary changes to your model

File: meshnets/model/model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (VictorGillioz) [reply meshnets/model/model.py:10]: The meshnets.model.x is an absolute import from the meshnets directory. This allows for example to easily and clearly import between subfolders in the project. I implemented it this way through the project but it can be changed to local imports instead if you think it is better.

For now I will change the "model" directory to "modules" and create the empty class files.

[the remark this answers is not in the document]
Comment (VictorGillioz) [reply meshnets/model/model.py]: > I already saw that they are not implemented. That was my fault. However when something is not implemented we shoudl replace the import with:
> 
> ```python
> class GraphEncoder:
>     pass
> ```

Where am I supposed to include this in general ? I gives pylint errors if the implemented classes are not coherent with the rest of the code (e.g. callable, correct number of arguments, class docstring). That seems very demanding for empty placeholder classes.

[the remark this answers is not in the document]
Comment (VictorGillioz) [reply meshnets/model/model.py]: Please let me know if the way I did it works for you @AugustoPeres !

**Review by AugustoPeres — APPROVED** [review]
Good job at addressing the most important comments!

Going to pre-aprove so that this can be merged.
Comment for later: I don't think we need a separate file for the encoder, decoder, processor. They are all neural network "layers". We can just have `meshenets/modules.py` or something. But we can figure out as we go

**Comment by VictorGillioz** [comment]
Regarding the custom Graph type, I think it emerges from the fact that the original work has a MultiGraph custom type. This is used in their case to have two set of edges in the same object. However, as we only need one edge set in this project the Data type from torch_geometric is more adapted and I will implement it from now on.
